proof/refine/ARM/Finalise_R.thy

```diff
@@ -2256,53 +2256,90 @@ lemma ntfn_bound_tcb_at':
-             irqs_masked_lift setBoundNotification_ct_not_inQ
-             untyped_ranges_zero_lift
-             | wpc | clarsimp simp: cteCaps_of_def)+
-  apply (clarsimp simp: o_def)
-  sorry (* unbindMaybeNotification_invs *) (*
-  apply safe[1]
-           defer 3
-           defer 7
-           apply (fold_subgoals (prefix))[8]
-           subgoal premises prems using prems by (auto simp: pred_tcb_at' valid_pspace'_def projectKOs valid_obj'_def valid_ntfn'_def
-                             ko_wp_at'_def
-                      elim!: obj_atE' valid_objsE' if_live_then_nonz_capE'
-                      split: option.splits ntfn.splits)
-   apply (rule delta_sym_refs, assumption)
-    apply (fold_subgoals (prefix))[2]
-    subgoal premises prems using prems by (fastforce simp: symreftype_inverse' ntfn_q_refs_of'_def
-                    split: ntfn.splits if_split_asm
-                     dest!: ko_at_state_refs_ofD')+
-  apply (rule delta_sym_refs, assumption)
-   apply (clarsimp split: if_split_asm)
-   apply (frule ko_at_state_refs_ofD', simp)
-  apply (clarsimp split: if_split_asm)
-   apply (frule_tac P="(=) (Some ntfnptr)" in ntfn_bound_tcb_at', simp_all add: valid_pspace'_def)[1]
-   subgoal by (fastforce simp: ntfn_q_refs_of'_def state_refs_of'_def tcb_ntfn_is_bound'_def
-                          tcb_st_refs_of'_def
-                   dest!: bound_tcb_at_state_refs_ofD'
-                   split: ntfn.splits thread_state.splits)
-  apply (frule ko_at_state_refs_ofD', simp)
-  done *)
+  apply (wpsimp wp: sbn'_valid_pspace'_inv sbn_sch_act' sbn_valid_queues
+                    valid_irq_node_lift irqs_masked_lift setBoundNotification_ct_not_inQ
+                    untyped_ranges_zero_lift
+                simp: cteCaps_of_def)+
+  by (auto simp: pred_tcb_at' valid_pspace'_def projectKOs valid_obj'_def
+                 valid_ntfn'_def ko_wp_at'_def live_ntfn'_def o_def
+          elim!: obj_atE' if_live_then_nonz_capE'
+          split: option.splits ntfn.splits)
 
 crunches getReplyTCB
   for inv[wp]: P
 
+lemma setNotification_invs':
+  "\<lbrace>invs'
+    and (\<lambda>s. live_ntfn' ntfn \<longrightarrow> ex_nonz_cap_to' ntfnPtr s)
+    and valid_ntfn' ntfn
+    and (\<lambda>s. sym_refs
+             (\<lambda>a. if a = ntfnPtr
+                   then refs_of_ntfn' ntfn
+                   else state_refs_of' s a)) \<rbrace>
+    setNotification ntfnPtr ntfn
+    \<lbrace>\<lambda>rv. invs'\<rbrace>"
+  apply (simp add:  invs'_def valid_state'_def)
+  apply (wpsimp wp: valid_pde_mappings_lift' untyped_ranges_zero_lift simp: cteCaps_of_def o_def)
+  done
+
+lemma setSchedContext_utr[wp]:
+  "\<lbrace>untyped_ranges_zero'\<rbrace> setSchedContext scPtr sc \<lbrace>\<lambda>rv. untyped_ranges_zero'\<rbrace>"
+  apply (simp add: cteCaps_of_def)
+  apply (rule hoare_pre, wp untyped_ranges_zero_lift)
+  apply (simp add: o_def)
+  done
+
+lemma schedContextUnbindNtfn_invs'[wp]:
```

Comment:
We'll need `schedContextUnbindNtfn_valid_objs'` in Ipc_R. Would you be able to factor that out into a separate lemma and use it in this proof?

Reply:
Done. Thanks for the suggestion, it even ended up improving the performance of the proof as well!